Здравствуйте, pull requests для hw1 был отправлен на ваш основной проект. Будут благодарен, если проверите его там.

Импортируем нужные библиотеки

In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import roc_auc_score, matthews_corrcoef
from sklearn.dummy import DummyClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline


Прочитаем данные и установим сид

In [ ]:
random_seed = 2026
df = pd.read_csv('/UCI_Credit_Card.csv')

X = df.drop(['ID', 'default.payment.next.month'], axis=1)
y = df['default.payment.next.month']

Разделим на тестовую и тренировочные выборки

In [18]:

cat_cols = ['LIMIT_BAL','SEX', 'EDUCATION', 'MARRIAGE', 'PAY_0', 'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6']
X_encoded = pd.get_dummies(X, columns=cat_cols, drop_first=True)

X_train, X_test, y_train, y_test = train_test_split(
    X_encoded, y, test_size=0.2, random_state=random_seed, stratify=y
)

print(X_train.shape, X_test.shape)

(24000, 161) (6000, 161)


В дз 1 были выбраны такие метрики как MCC и AUC ROC. Их и будем использовать для оценивания качества моделей.  
Для константного предсказания будем использовать самый частые класс.

In [16]:
dummy = DummyClassifier(strategy='most_frequent', random_state=random_seed)
dummy.fit(X_train, y_train)
y_pred_dummy = dummy.predict(X_test)
y_prob_dummy = dummy.predict_proba(X_test)[:, 1]

print('Dummy MCC: ', matthews_corrcoef(y_test, y_pred_dummy))
print('Dummy AUC: ', roc_auc_score(y_test, y_prob_dummy))

Dummy MCC:  0.0
Dummy AUC:  0.5


Для простой модели используем логическую регрессию с такими же метриками. Все парметры, кроме сида, будут стандартными для максимальной простоты.

In [19]:
pipe = make_pipeline(
    StandardScaler(),
    LogisticRegression(random_state=random_seed)
)

pipe.fit(X_train, y_train)
y_pred = pipe.predict(X_test)
y_prob = pipe.predict_proba(X_test)[:, 1]

print('LR MCC: ', matthews_corrcoef(y_test, y_pred))
print('LR AUC: ', roc_auc_score(y_test, y_prob))

LR MCC:  0.4007858052464442
LR AUC:  0.7718534427359403


При оценке параметров можно сказать, что модель рпоказывает значительно качество в сравнение с констаным предсказанием, AUC = 0.77 говорит о том, что модель может выдавать правдивые результаты на основе своих данных.